File: generate_apple_secret.py
import jwt
import time
import sys

def generate_client_secret(key_file, team_id, client_id, key_id):
    with open(key_file, 'r') as f:
        private_key = f.read()

    headers = {
        'kid': key_id,
        'alg': 'ES256'
    }

    payload = {
        'iss': team_id,
        'iat': int(time.time()),
        'exp': int(time.time()) + 15777000, # 6 months in seconds
        'aud': 'https://appleid.apple.com',
        'sub': client_id,
    }

    client_secret = jwt.encode(
        payload,
        private_key,
        algorithm='ES256',
        headers=headers
    )

    return client_secret

if __name__ == "__main__":
    print("-" * 50)
    print("Apple Client Secret Generator")
    print("-" * 50)
    
    # Configuration
    TEAM_ID = "KXVAN4J7F3" 
    # Using the Services ID as the subject for the JWT
    CLIENT_ID = "com.yuna.yoga.bauti.signin" 
    KEY_ID = "UZ3W6BA49U"
    KEY_FILE = "AuthKey_UZ3W6BA49U.p8"

    try:
        secret = generate_client_secret(KEY_FILE, TEAM_ID, CLIENT_ID, KEY_ID)
        print("\n✅ Generated Client Secret JWT (Copy this to Supabase):")
        print("-" * 20)
        print(secret)
        print("-" * 20)
    except FileNotFoundError:
        print(f"\n❌ Error: '{KEY_FILE}' not found.")
    except Exception as e:
        print(f"\n❌ Error: {e}")

File: generate_apple_secret.js
const crypto = require('crypto');
const fs = require('fs');

function generateClientSecret(keyFile, teamId, clientId, keyId) {
    const privateKey = fs.readFileSync(keyFile, 'utf8');

    const header = {
        alg: 'ES256',
        kid: keyId,
        typ: 'JWT'
    };

    const now = Math.floor(Date.now() / 1000);
    const payload = {
        iss: teamId,
        iat: now,
        exp: now + 15777000, // 6 months
        aud: 'https://appleid.apple.com',
        sub: clientId
    };

    const base64UrlEncode = (str) => {
        return Buffer.from(JSON.stringify(str))
            .toString('base64')
            .replace(/\+/g, '-')
            .replace(/\//g, '_')
            .replace(/=/g, '');
    };

    const encodedHeader = base64UrlEncode(header);
    const encodedPayload = base64UrlEncode(payload);

    const data = `${encodedHeader}.${encodedPayload}`;
    const signer = crypto.createSign('RSA-SHA256'); // Wait, ES256 is ECDSA with P-256 and SHA-256

    // Correct way for ES256 in Node crypto:
    const sign = crypto.sign("sha256", Buffer.from(data), {
        key: privateKey,
        dsaEncoding: 'ieee-p1363'
    }).toString('base64').replace(/\+/g, '-').replace(/\//g, '_').replace(/=/g, '');

    return `${data}.${sign}`;
}

const TEAM_ID = "KXVAN4J7F3";
const CLIENT_ID = "com.yuna.yoga.bauti"; // Use Bundle ID for mobile apps
const KEY_ID = "CYMA73GRPV";
const KEY_FILE = "AuthKey_CYMA73GRPV.p8";

try {
    const secret = generateClientSecret(KEY_FILE, TEAM_ID, CLIENT_ID, KEY_ID);
    console.log("\n✅ Generated Client Secret JWT:");
    console.log("--------------------");
    console.log(secret);
    console.log("--------------------");
} catch (e) {
    console.error("\n❌ Error:", e.message);
}
